src/pygeoprocessing/geoprocessing.py

```diff
@@ -1408,373 +1421,357 @@ def zonal_statistics(
-                        dict(zip(*numpy.unique(
-                            masked_clipped_block, return_counts=True))))
-    unset_fids = aggregate_layer_fid_set.difference(aggregate_stats)
-    LOGGER.debug(
-        "unset_fids: %s of %s ", len(unset_fids),
-        len(aggregate_layer_fid_set))
-    clipped_gt = numpy.array(
-        clipped_raster.GetGeoTransform(), dtype=numpy.float32)
-    LOGGER.debug("gt %s for %s", clipped_gt, base_raster_path_band)
-    for unset_fid in unset_fids:
-        unset_feat = aggregate_layer.GetFeature(unset_fid)
-        unset_geom_ref = unset_feat.GetGeometryRef()
-        if unset_geom_ref is None:
-            LOGGER.warning(
-                f'no geometry in {aggregate_vector_path} FID: {unset_fid}')
-            continue
-
-        unset_geom_envelope = list(unset_geom_ref.GetEnvelope())
-        unset_geom_ref = None
-        unset_feat = None
-        if clipped_gt[1] < 0:
-            unset_geom_envelope[0], unset_geom_envelope[1] = (
-                unset_geom_envelope[1], unset_geom_envelope[0])
-        if clipped_gt[5] < 0:
-            unset_geom_envelope[2], unset_geom_envelope[3] = (
-                unset_geom_envelope[3], unset_geom_envelope[2])
-
-        xoff = int((unset_geom_envelope[0] - clipped_gt[0]) / clipped_gt[1])
-        yoff = int((unset_geom_envelope[2] - clipped_gt[3]) / clipped_gt[5])
-        win_xsize = int(numpy.ceil(
-            (unset_geom_envelope[1] - clipped_gt[0]) /
-            clipped_gt[1])) - xoff
-        win_ysize = int(numpy.ceil(
-            (unset_geom_envelope[3] - clipped_gt[3]) /
-            clipped_gt[5])) - yoff
-
-        # clamp offset to the side of the raster if it's negative
-        if xoff < 0:
-            win_xsize += xoff
-            xoff = 0
-        if yoff < 0:
-            win_ysize += yoff
-            yoff = 0
-
-        # clamp the window to the side of the raster if too big
-        if xoff+win_xsize > clipped_band.XSize:
-            win_xsize = clipped_band.XSize-xoff
-        if yoff+win_ysize > clipped_band.YSize:
-            win_ysize = clipped_band.YSize-yoff
-
-        if win_xsize <= 0 or win_ysize <= 0:
-            continue
+        disjoint_fid_sets = [valid_fid_set]
+
+    # Rasterize each disjoint polygon set onto its own raster layer
+    fid_nodata = -1
+    fid_raster_paths = []
+    for i, disjoint_fid_set in enumerate(disjoint_fid_sets):
+        fid_raster_path = os.path.join(temp_working_dir, f'fid_set_{i}.tif')
+        fid_set_str = ", ".join(str(fid) for fid in disjoint_fid_set)
+        gdal.Rasterize(
+            destNameOrDestDS=fid_raster_path,
+            srcDS=target_vector,
+            allTouched=False,
+            attribute=fid_field_name,
+            noData=fid_nodata,
+            outputBounds=bbox_intersection,
+            xRes=raster_info['pixel_size'][0],
+            yRes=raster_info['pixel_size'][1],
+            format='GTIFF',
+            outputType=gdal.GDT_Int32,
```

Comment:
Sure, that will work if I change the nodata value to `numpy.iinfo(numpy.uint16).max`

Reply:
The optimization you suggest would be really neat to have, but it feels like it could be a whole separate PR. Maybe that could be abstracted away in a separate pygeoprocessing function?

Reply:
>The optimization you suggest would be really neat to have, but it feels like it could be a whole separate PR. Maybe that could be abstracted away in a separate pygeoprocessing function?

Eh, I'm not convinced it would really be worth it for the extra iteration it would impose given that [`max` is O(n)](https://wiki.python.org/moin/TimeComplexity) and the disjoint sets can be large.  In any case, let's leave it out of this PR.